File: rob.java
//Constant space solution
class Solution {
    public int rob(int[] nums) {
        //Tc: O(n)  Sc: O(1)
        if(nums == null || nums.length == 0) return 0;
        int n = nums.length;
        int dp[] = new int[n];
        if(n<2) return nums[0];
        int prev = nums[0];
        int curr = Math.max(nums[0], nums[1]);
        for(int i = 2; i < nums.length; i++)
        {
            int temp = curr;
            curr = Math.max(prev+nums[i], curr);
            prev = temp;
        }
        return curr;
//DP
        /*
        //Tc: O(n)  Sc: O(n)
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0],nums[1]);
        for(int i = 2; i < n ; i++)
        {
            dp [i] = Math.max(dp[i-1], nums[i]+ dp[i-2]);
        }
        return dp[n-1];
        */
    }
}

//Exhaustive Approach
/*
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;

        return helper(nums, 0, 0);
      
    }

    private int helper(int[] nums, int index, int amount)
    {
        //base case
        if(index >= nums.length) return amount;
        //logic
        //If we want to select the num
        int case1 = helper(nums, index+2, amount+nums[index]);
        // Don't want to select the num
        int case2 = helper(nums, index+1, amount);

        return Math.max(case1, case2);
    }
}
*/

File: Rob.py
'''
    Time Complexity:
        O(n) (where n is the length of the given array)

    Space Complexity:
        O(n) (where n is the length of the given array)

    Did this code successfully run on LeetCode?:
        Yes

    Problems faced while coding this:
        None

    Approach:
        Top Down Memoized DP approach.
        Traverse the array.
        In each traversal, either choose this and skip the next.
        Or choose next.
        Return the max of those two.
'''
class Solution:
    def __init__(self):
        self.dp = {}
        self.nums = []

    def rob(self, nums: List[int]) -> int:
        self.nums = nums
        return self.find_max(0)

    def find_max(self, idx):
        if idx in self.dp:
            return self.dp[idx]

        if idx >= len(self.nums):
            return 0

        select = self.nums[idx] + self.find_max(idx + 2)
        ignore = self.find_max(idx + 1)
        self.dp[idx] = max(select, ignore)
        return self.dp[idx]
